### ISSUE TITLE
☂️ Dart 3 framework changes

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#118837]

**Comment by goderbauer** [comment]
Belated null safety dividend: Thanks to Dart 3 removing support for unsafe code we were just able to remove almost 5,000 lines of now unnecessary null comparisons from the `flutter/flutter` and `flutter/engine` code base. Pretty nice clean-up (see PRs linked to this issue above).

**Comment by goderbauer** [comment]
cc @mit-mit Let me know if anything else Dart 3 related comes to mind that we should prepare for in Flutter.

**Comment by mit-mit** [comment]
The only thing I can think of are the CLI messages, but that's already tracked in https://github.com/flutter/flutter/issues/115803

**Comment by goderbauer** [comment]
Top-level comment updated with some work that needs to happen for class modifiers.

**Comment by goderbauer** [comment]
Flag flips for the Dart 3 features happened upstream in the Dart SDK with these CLs:

* https://dart-review.googlesource.com/c/sdk/+/285080
* https://dart-review.googlesource.com/c/sdk/+/286344

I'll prepare PRs to bump our Dart SDK lower constraint to 3.0 to enable these features.

**Comment by goderbauer** [comment]
Web engine breaks when bumping it to Dart 3.0, details in https://github.com/flutter/flutter/issues/122724. For now, I am excluding the web from the version bump to unblock framework. Not sure yet who can help with debugging this.

**Comment by goderbauer** [comment]
Web compiler bug is fixed. Dart 3.0 is fully enabled for the engine repository now and some class modifiers have been applied to the engine code base to prove that the new feature can be used (https://github.com/flutter/engine/pull/40328). 

The work to upgrade the flutter/flutter repository to 3.0 is currently blocked on waiting for https://github.com/dart-lang/sdk/commit/7cc2553fa1f2c4fb34dee275c4fe281c89a90cdf to roll in.

cc @jakemac53 @itsjustkevin

**Comment by goderbauer** [comment]
https://github.com/dart-lang/sdk/commit/7cc2553fa1f2c4fb34dee275c4fe281c89a90cdf is tagged with 3.0.0-345.0.dev. Once that version (or something newer) has rolled into Flutter, work can continue here.

**Comment by goderbauer** [comment]
While bumping the framework to 3.0, an issue related to const constructors surfaced: https://github.com/dart-lang/sdk/issues/51800

**Comment by goderbauer** [comment]
And another web compiler issue: https://github.com/dart-lang/sdk/issues/51802

**Comment by goderbauer** [comment]
Tracking issue to clean up the work arounds we had to apply: https://github.com/dart-lang/sdk/issues/51813

**Comment by goderbauer** [comment]
The flutter/flutter repository is now also fully Dart 3 enabled. There are a few follow-ups (see checklist in first post and https://github.com/dart-lang/sdk/issues/51813), but the majority of the work is done here.

Of the remaining work, getting https://github.com/dart-lang/sdk/issues/51823 resolved is probably the most critical, so we can remove the work around for it and avoid future surprises.
